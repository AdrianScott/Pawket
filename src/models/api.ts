export interface ApiResponse {
  error?: string;
  success?: boolean;
}

export interface GetParentPuzzleRequest {
  parentCoinId: string;
}

export interface GetParentPuzzleResponse {
  amount: number;
  parentCoinId: string;
  parentParentCoinId: string;
  puzzleReveal: string;
}
