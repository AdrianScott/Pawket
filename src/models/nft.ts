import { NftDetail } from "@/services/crypto/receive";
import { OriginCoin } from "./wallet";

export interface NftItemAttribute {
  trait_type: string;
  value: string;
  min_value: string;
  max_value: string;
}

export interface NftCollectionAttribute {
  type: string;
  value: string;
}

export interface NftCollection {
  name: string;
  id: string;
  attributes: NftCollectionAttribute[];
}

interface DownloadedNftInfo {
  metadata?: NftOffChainMetadata;
  matchHash?: boolean;
  status: "Ready" | "NoMetadata" | "Downloading" | "Processed";
}

export type DonwloadedNftCollection = { [name: string]: DownloadedNftInfo };

export interface NftOffChainMetadata {
  format: "CHIP-0007";
  name: string;
  description: string;
  minting_tool: string;
  sensitive_content: boolean;
  series_number: number;
  series_total: number;
  attributes: NftItemAttribute[];
  collection: NftCollection;
  levels: string[];
  stats: string[];
  data: unknown;
}

export interface NftCoinAnalysisResult {
  singletonModHash: string;
  launcherId: string;
  launcherPuzzleHash: string;
  nftStateModHash: string;
  metadataUpdaterPuzzleHash: string;
  p2InnerPuzzle: string;
  hintPuzzle: string;
  nftOwnershipModHash: string,
  previousOwner: string;
  didOwner: string;
  p2Owner: string;
  royaltyAddress: string;
  tradePricePercentage: number;
  rawMetadata: string;
  metadata: NftMetadataValues;
  coin: OriginCoin;
  nextCoinName?: string;
  updaterInSolution?: boolean;
}

export type NftDataKey = "imageUri" | "imageHash" | "metadataUri" | "metadataHash" | "licenseUri" | "licenseHash" | "serialNumber" | "serialTotal"
  // CNS related keys
  | "address" | "name" | "text";
export type NftMetadataValues = {
  imageUri: string | string[] | undefined;
  imageHash: string | undefined;
  metadataUri: string | string[] | undefined;
  metadataHash: string | undefined;
  licenseUri: string | string[] | undefined;
  licenseHash: string | undefined;
  serialNumber: string | undefined;
  serialTotal: string | undefined;
  // CNS related keys
  address?: string;
  name?: string;
  text?: string;
};
export type NftMetadataKeys = { [key in NftDataKey]: string; };