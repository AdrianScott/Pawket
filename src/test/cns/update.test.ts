import { getTestAccount } from "../utility";
import { SymbolCoins } from "@/services/transfer/transfer";
import { analyzeNftCoin, generateUpdatedNftBundle } from "@/services/coin/nft";
import puzzle from "@/services/crypto/puzzle";
import { GetParentPuzzleResponse } from "@/models/api";
import { Instance } from "@/services/util/instance";
import { getAccountAddressDetails } from "@/services/util/account";

import { knownCoins } from "./cns.test.data";
import { convertToOriginCoin, NetworkContext } from "@/services/coin/coinUtility";

import cnscoin1 from "../cases/cnscoin1.json"
import { assertSpendbundle } from "@/services/spendbundle/validator";
import { signSpendBundle } from "@/services/spendbundle";

const net: NetworkContext = {
  prefix: "xch",
  symbol: "XCH",
  chainId: "ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb",
  api: localPuzzleApiCall,
}
function xchPrefix() { return "xch"; }
function xchSymbol() { return "XCH"; }
function chainId() { return "ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb"; }
function tokenInfo() { return {}; }

beforeAll(async () => {
  await Instance.init();
})

export async function testUpdateCns(fee: bigint): Promise<void> {
  const hintPuzzle = "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10";
  const change_hex = "0x0eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155";
  const update_hex = "0xd26c36cfd99da03a18a7d47dddd7beb968ff63bd7d3ccc45205fadb6958a571d";
  const nftCoin = convertToOriginCoin({
    "amount": 1,
    "parent_coin_info": "0x76f8a0a38e04a54a3f08a65bc5860efdd6a52b678060c9747329248a8b1ad0dd",
    "puzzle_hash": "0xe1d6b7f6088ec67185c53524be7ceeb02df71ed99cdd4f5af1e120c86a3091e4"
  });

  const analysis = await analyzeNftCoin(cnscoin1.puzzle_reveal, hintPuzzle, nftCoin, cnscoin1.solution);
  if (analysis == null) fail("null analysis");
  expect(analysis).toMatchSnapshot("analysis");

  const account = getTestAccount("55c335b84240f5a8c93b963e7ca5b868e0308974e09f751c7e5668964478008f");
  account.addressRetrievalCount = 5;

  const changeAddress = puzzle.getAddressFromPuzzleHash(change_hex, xchPrefix());

  const tokenPuzzles = await getAccountAddressDetails(account, [], tokenInfo(), xchPrefix(), xchSymbol(), undefined, "cat_v1");
  const availcoins: SymbolCoins = {
    [xchSymbol()]: [
      {
        "parent_coin_info": "0xc2d7aa805ebbfdfdd5760294b74ddd63c76cae14da817cd27e40ea2c15f18298",
        "puzzle_hash": "0x0eb720d9195ffe59684b62b12d54791be7ad3bb6207f5eb92e0e1b40ecbc1155",
        "amount": 21n
      },
    ]
  };

  const ubundle = await generateUpdatedNftBundle(
    changeAddress, fee, nftCoin, analysis, "CNS", "address", update_hex, availcoins, tokenPuzzles, net);

  const finalCoin = ubundle.coin_spends[0];
  const finalAnalysis = await analyzeNftCoin(finalCoin.puzzle_reveal, analysis.p2Owner, finalCoin.coin, finalCoin.solution);
  if (finalAnalysis == null) fail("null finalAnalysis");
  expect(finalAnalysis).toMatchSnapshot("finalAnalysis");

  const spendBundle = await signSpendBundle(ubundle, tokenPuzzles, net.chainId);
  await assertSpendbundle(spendBundle, net.chainId);
  expect(spendBundle).toMatchSnapshot("spendbundle");
}

test('Update CNS', async () => {
  await testUpdateCns(0n);
});

async function localPuzzleApiCall(parentCoinId: string): Promise<GetParentPuzzleResponse | undefined> {
  const resp = knownCoins.find(_ => _.parentCoinId == parentCoinId);
  return resp;
}
