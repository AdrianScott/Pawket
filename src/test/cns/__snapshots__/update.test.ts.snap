// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update CNS: analysis 1`] = `
Object {
  "cnsAddress": "55874c54ec86a88e22d3ad4f2ab6422d0c0b98d2dbe997dd21f40593d06c8267",
  "cnsExpiry": "1669907218",
  "cnsName": "icertest.xch",
  "coin": Object {
    "amount": 1n,
    "parent_coin_info": "0x76f8a0a38e04a54a3f08a65bc5860efdd6a52b678060c9747329248a8b1ad0dd",
    "puzzle_hash": "0xe1d6b7f6088ec67185c53524be7ceeb02df71ed99cdd4f5af1e120c86a3091e4",
  },
  "didOwner": "",
  "hintPuzzle": "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
  "launcherId": "0x1dd352dccae28edfcba35fbcdcab358af8d00979899dea9b8c79a9f29a1fd450",
  "launcherPuzzleHash": "0xeff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9",
  "metadata": Object {
    "address": "55874c54ec86a88e22d3ad4f2ab6422d0c0b98d2dbe997dd21f40593d06c8267",
    "expiry": "1669907218",
    "imageHash": "f7ecc9f2bc09682e67d3a81445223a2dfa5814cddd33b6d6b894f296b446bc18",
    "imageUri": "https://dweb.link/ipfs/QmevcSM1CaDAuBq5kcUc3eKZAKWmeYb8tWkmtRgNhEV1SX?filename=icertest.xch.jpg",
    "licenseHash": "",
    "licenseUri": "",
    "metadataHash": "7060452f409047d878babfffa77c6df74a15324bacee9a78bb1901cebb8b91e3",
    "metadataUri": "https://dweb.link/ipfs/QmPGmorPHXa3rjUbv63dNE41xSoMkUqUHDfH8jxiayGZxE?filename=icertest.xchmetadata.json",
    "name": "icertest.xch",
    "serialNumber": "01",
    "serialTotal": "01",
  },
  "metadataUpdaterPuzzleHash": "0x36a5a899b39a95f0fc7421f65f2e79562684bff71434c1799f33a57354649eee",
  "nextCoinName": "0x24f34fcbebf8c207f34a852485380ce94f365d97e3679e55429b0b0e0b93fd73",
  "nftOwnershipModHash": "0xc5abea79afaa001b5427dfa0c8cf42ca6f38f5841b78f9b3c252733eb2de2726",
  "nftStateModHash": "0xa04d9f57764f54a43e4030befb4d80026e870519aaa66334aef8304f5d0393c2",
  "p2InnerPuzzle": "(a (q 2 (q 2 (i 11 (q 2 (i (= 5 (point_add 11 (pubkey_for_exp (sha256 11 (a 6 (c 2 (c 23 ()))))))) (q 2 23 47) (q 8)) 1) (q 4 (c 4 (c 5 (c (a 6 (c 2 (c 23 ()))) ()))) (a 23 47))) 1) (c (q 50 2 (i (l 5) (q 11 (q . 2) (a 6 (c 2 (c 9 ()))) (a 6 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1)) (c (q . 0xa9cb314c3e76c877d5ad4188b81d1faac3d8835e0f50a13d20bbd935966fdd9b1b38c7d910f0cf926ab55615ea3ab643) 1))",
  "p2Owner": "7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
  "previousOwner": "()",
  "rawMetadata": "((117 \\"https://dweb.link/ipfs/QmevcSM1CaDAuBq5kcUc3eKZAKWmeYb8tWkmtRgNhEV1SX?filename=icertest.xch.jpg\\") (104 . 0xf7ecc9f2bc09682e67d3a81445223a2dfa5814cddd33b6d6b894f296b446bc18) (28021 \\"https://dweb.link/ipfs/QmPGmorPHXa3rjUbv63dNE41xSoMkUqUHDfH8jxiayGZxE?filename=icertest.xchmetadata.json\\") (28008 . 0x7060452f409047d878babfffa77c6df74a15324bacee9a78bb1901cebb8b91e3) (27765) (27752) (29550 . 1) (29556 . 1) (25976 . 0x6388c312) (24932 . 0x55874c54ec86a88e22d3ad4f2ab6422d0c0b98d2dbe997dd21f40593d06c8267) (28269 . \\"icertest.xch\\"))",
  "royaltyAddress": "0x5662b49a357db4f05c2c141452b72fb91e7ec286e9b47d6c287210c63ae5cd3e",
  "singletonModHash": "0x7faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9f",
  "tradePricePercentage": 500,
}
`;

exports[`Update CNS: finalAnalysis 1`] = `
Object {
  "cnsAddress": "d26c36cfd99da03a18a7d47dddd7beb968ff63bd7d3ccc45205fadb6958a571d",
  "cnsExpiry": "1669907218",
  "cnsName": "icertest.xch",
  "coin": Object {
    "amount": 1n,
    "parent_coin_info": "0x76f8a0a38e04a54a3f08a65bc5860efdd6a52b678060c9747329248a8b1ad0dd",
    "puzzle_hash": "0xe1d6b7f6088ec67185c53524be7ceeb02df71ed99cdd4f5af1e120c86a3091e4",
  },
  "didOwner": "()",
  "hintPuzzle": "0x7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
  "launcherId": "0x1dd352dccae28edfcba35fbcdcab358af8d00979899dea9b8c79a9f29a1fd450",
  "launcherPuzzleHash": "0xeff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9",
  "metadata": Object {
    "address": "d26c36cfd99da03a18a7d47dddd7beb968ff63bd7d3ccc45205fadb6958a571d",
    "expiry": "1669907218",
    "imageHash": "f7ecc9f2bc09682e67d3a81445223a2dfa5814cddd33b6d6b894f296b446bc18",
    "imageUri": "https://dweb.link/ipfs/QmevcSM1CaDAuBq5kcUc3eKZAKWmeYb8tWkmtRgNhEV1SX?filename=icertest.xch.jpg",
    "licenseHash": "",
    "licenseUri": "",
    "metadataHash": "7060452f409047d878babfffa77c6df74a15324bacee9a78bb1901cebb8b91e3",
    "metadataUri": "https://dweb.link/ipfs/QmPGmorPHXa3rjUbv63dNE41xSoMkUqUHDfH8jxiayGZxE?filename=icertest.xchmetadata.json",
    "name": "icertest.xch",
    "serialNumber": "01",
    "serialTotal": "01",
  },
  "metadataUpdaterPuzzleHash": "0x36a5a899b39a95f0fc7421f65f2e79562684bff71434c1799f33a57354649eee",
  "nextCoinName": "0xd19259f275bb4dafdc40a671899b8227fe15466bab85748cd3999c0c25de0cb0",
  "nftOwnershipModHash": "0xc5abea79afaa001b5427dfa0c8cf42ca6f38f5841b78f9b3c252733eb2de2726",
  "nftStateModHash": "0xa04d9f57764f54a43e4030befb4d80026e870519aaa66334aef8304f5d0393c2",
  "p2InnerPuzzle": "(a (q 2 (q 2 (i 11 (q 2 (i (= 5 (point_add 11 (pubkey_for_exp (sha256 11 (a 6 (c 2 (c 23 ()))))))) (q 2 23 47) (q 8)) 1) (q 4 (c 4 (c 5 (c (a 6 (c 2 (c 23 ()))) ()))) (a 23 47))) 1) (c (q 50 2 (i (l 5) (q 11 (q . 2) (a 6 (c 2 (c 9 ()))) (a 6 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1)) (c (q . 0x92c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1d) 1))",
  "p2Owner": "7ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10",
  "previousOwner": "()",
  "rawMetadata": "((117 \\"https://dweb.link/ipfs/QmevcSM1CaDAuBq5kcUc3eKZAKWmeYb8tWkmtRgNhEV1SX?filename=icertest.xch.jpg\\") (104 . 0xf7ecc9f2bc09682e67d3a81445223a2dfa5814cddd33b6d6b894f296b446bc18) (28021 \\"https://dweb.link/ipfs/QmPGmorPHXa3rjUbv63dNE41xSoMkUqUHDfH8jxiayGZxE?filename=icertest.xchmetadata.json\\") (28008 . 0x7060452f409047d878babfffa77c6df74a15324bacee9a78bb1901cebb8b91e3) (27765) (27752) (29550 . 1) (29556 . 1) (25976 . 0x6388c312) (24932 . 0xd26c36cfd99da03a18a7d47dddd7beb968ff63bd7d3ccc45205fadb6958a571d) (28269 . \\"icertest.xch\\"))",
  "royaltyAddress": "0x5662b49a357db4f05c2c141452b72fb91e7ec286e9b47d6c287210c63ae5cd3e",
  "singletonModHash": "0x7faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9f",
  "tradePricePercentage": 500,
  "updaterInSolution": true,
}
`;

exports[`Update CNS: spendbundle 1`] = `
Object {
  "aggregated_signature": "0x8cce194259562d76c87e3ecac9d3da4cb2c43f3e2c45c62da3d568bd31f93d7a14b20cb8227842ce2230c1ac27629392094a128d093037262175aa545c402a69b2b4438eb805c3afe041f8414b908082661e8713ce76663eed25ce94f42c0619",
  "coin_spends": Array [
    Object {
      "coin": Object {
        "amount": 1n,
        "parent_coin_info": "0x76f8a0a38e04a54a3f08a65bc5860efdd6a52b678060c9747329248a8b1ad0dd",
        "puzzle_hash": "0xe1d6b7f6088ec67185c53524be7ceeb02df71ed99cdd4f5af1e120c86a3091e4",
      },
      "puzzle_reveal": "0xff02ffff01ff02ffff01ff02ffff03ffff18ff2fff3480ffff01ff04ffff04ff20ffff04ff2fff808080ffff04ffff02ff3effff04ff02ffff04ff05ffff04ffff02ff2affff04ff02ffff04ff27ffff04ffff02ffff03ff77ffff01ff02ff36ffff04ff02ffff04ff09ffff04ff57ffff04ffff02ff2effff04ff02ffff04ff05ff80808080ff808080808080ffff011d80ff0180ffff04ffff02ffff03ff77ffff0181b7ffff015780ff0180ff808080808080ffff04ff77ff808080808080ffff02ff3affff04ff02ffff04ff05ffff04ffff02ff0bff5f80ffff01ff8080808080808080ffff01ff088080ff0180ffff04ffff01ffffffff4947ff0233ffff0401ff0102ffffff20ff02ffff03ff05ffff01ff02ff32ffff04ff02ffff04ff0dffff04ffff0bff3cffff0bff34ff2480ffff0bff3cffff0bff3cffff0bff34ff2c80ff0980ffff0bff3cff0bffff0bff34ff8080808080ff8080808080ffff010b80ff0180ffff02ffff03ffff22ffff09ffff0dff0580ff2280ffff09ffff0dff0b80ff2280ffff15ff17ffff0181ff8080ffff01ff0bff05ff0bff1780ffff01ff088080ff0180ff02ffff03ff0bffff01ff02ffff03ffff02ff26ffff04ff02ffff04ff13ff80808080ffff01ff02ffff03ffff20ff1780ffff01ff02ffff03ffff09ff81b3ffff01818f80ffff01ff02ff3affff04ff02ffff04ff05ffff04ff1bffff04ff34ff808080808080ffff01ff04ffff04ff23ffff04ffff02ff36ffff04ff02ffff04ff09ffff04ff53ffff04ffff02ff2effff04ff02ffff04ff05ff80808080ff808080808080ff738080ffff02ff3affff04ff02ffff04ff05ffff04ff1bffff04ff34ff8080808080808080ff0180ffff01ff088080ff0180ffff01ff04ff13ffff02ff3affff04ff02ffff04ff05ffff04ff1bffff04ff17ff8080808080808080ff0180ffff01ff02ffff03ff17ff80ffff01ff088080ff018080ff0180ffffff02ffff03ffff09ff09ff3880ffff01ff02ffff03ffff18ff2dffff010180ffff01ff0101ff8080ff0180ff8080ff0180ff0bff3cffff0bff34ff2880ffff0bff3cffff0bff3cffff0bff34ff2c80ff0580ffff0bff3cffff02ff32ffff04ff02ffff04ff07ffff04ffff0bff34ff3480ff8080808080ffff0bff34ff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff02ffff03ffff21ff17ffff09ff0bff158080ffff01ff04ff30ffff04ff0bff808080ffff01ff088080ff0180ff018080ffff04ffff01ffa07faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9fffa01dd352dccae28edfcba35fbcdcab358af8d00979899dea9b8c79a9f29a1fd450a0eff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9ffff04ffff01ff02ffff01ff02ffff01ff02ff3effff04ff02ffff04ff05ffff04ffff02ff2fff5f80ffff04ff80ffff04ffff04ffff04ff0bffff04ff17ff808080ffff01ff808080ffff01ff8080808080808080ffff04ffff01ffffff0233ff04ff0101ffff02ff02ffff03ff05ffff01ff02ff1affff04ff02ffff04ff0dffff04ffff0bff12ffff0bff2cff1480ffff0bff12ffff0bff12ffff0bff2cff3c80ff0980ffff0bff12ff0bffff0bff2cff8080808080ff8080808080ffff010b80ff0180ffff0bff12ffff0bff2cff1080ffff0bff12ffff0bff12ffff0bff2cff3c80ff0580ffff0bff12ffff02ff1affff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff02ffff03ff0bffff01ff02ffff03ffff09ff23ff1880ffff01ff02ffff03ffff18ff81b3ff2c80ffff01ff02ffff03ffff20ff1780ffff01ff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff33ffff04ff2fffff04ff5fff8080808080808080ffff01ff088080ff0180ffff01ff04ff13ffff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff17ffff04ff2fffff04ff5fff80808080808080808080ff0180ffff01ff02ffff03ffff09ff23ffff0181e880ffff01ff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff17ffff04ffff02ffff03ffff22ffff09ffff02ff2effff04ff02ffff04ff53ff80808080ff82014f80ffff20ff5f8080ffff01ff02ff53ffff04ff818fffff04ff82014fffff04ff81b3ff8080808080ffff01ff088080ff0180ffff04ff2cff8080808080808080ffff01ff04ff13ffff02ff3effff04ff02ffff04ff05ffff04ff1bffff04ff17ffff04ff2fffff04ff5fff80808080808080808080ff018080ff0180ffff01ff04ffff04ff18ffff04ffff02ff16ffff04ff02ffff04ff05ffff04ff27ffff04ffff0bff2cff82014f80ffff04ffff02ff2effff04ff02ffff04ff818fff80808080ffff04ffff0bff2cff0580ff8080808080808080ff378080ff81af8080ff0180ff018080ffff04ffff01a0a04d9f57764f54a43e4030befb4d80026e870519aaa66334aef8304f5d0393c2ffff04ffff01ffff75ffc05f68747470733a2f2f647765622e6c696e6b2f697066732f516d657663534d3143614441754271356b63556333654b5a414b576d6559623874576b6d7452674e6845563153583f66696c656e616d653d69636572746573742e7863682e6a706780ffff68a0f7ecc9f2bc09682e67d3a81445223a2dfa5814cddd33b6d6b894f296b446bc18ffff826d75ffc06868747470733a2f2f647765622e6c696e6b2f697066732f516d50476d6f725048586133726a5562763633644e45343178536f4d6b55715548446648386a78696179475a78453f66696c656e616d653d69636572746573742e7863686d657461646174612e6a736f6e80ffff826d68a07060452f409047d878babfffa77c6df74a15324bacee9a78bb1901cebb8b91e3ffff826c7580ffff826c6880ffff82736e01ffff82737401ffff826578846388c312ffff826164a055874c54ec86a88e22d3ad4f2ab6422d0c0b98d2dbe997dd21f40593d06c8267ffff826e6d8c69636572746573742e78636880ffff04ffff01a036a5a899b39a95f0fc7421f65f2e79562684bff71434c1799f33a57354649eeeffff04ffff01ff02ffff01ff02ffff01ff02ff26ffff04ff02ffff04ff05ffff04ff17ffff04ff0bffff04ffff02ff2fff5f80ff80808080808080ffff04ffff01ffffff82ad4cff0233ffff3e04ff81f601ffffff0102ffff02ffff03ff05ffff01ff02ff2affff04ff02ffff04ff0dffff04ffff0bff32ffff0bff3cff3480ffff0bff32ffff0bff32ffff0bff3cff2280ff0980ffff0bff32ff0bffff0bff3cff8080808080ff8080808080ffff010b80ff0180ff04ffff04ff38ffff04ffff02ff36ffff04ff02ffff04ff05ffff04ff27ffff04ffff02ff2effff04ff02ffff04ffff02ffff03ff81afffff0181afffff010b80ff0180ff80808080ffff04ffff0bff3cff4f80ffff04ffff0bff3cff0580ff8080808080808080ff378080ff82016f80ffffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff2fffff01ff80ff808080808080808080ff0bff32ffff0bff3cff2880ffff0bff32ffff0bff32ffff0bff3cff2280ff0580ffff0bff32ffff02ff2affff04ff02ffff04ff07ffff04ffff0bff3cff3c80ff8080808080ffff0bff3cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff02ffff03ff5fffff01ff02ffff03ffff09ff82011fff3880ffff01ff02ffff03ffff09ffff18ff82059f80ff3c80ffff01ff02ffff03ffff20ff81bf80ffff01ff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff82019fffff04ff82017fff80808080808080808080ffff01ff088080ff0180ffff01ff04ff819fffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ff82017fff808080808080808080808080ff0180ffff01ff02ffff03ffff09ff82011fff2c80ffff01ff02ffff03ffff20ff82017f80ffff01ff04ffff04ff24ffff04ffff0eff10ffff02ff2effff04ff02ffff04ff82019fff8080808080ff808080ffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ffff02ff0bffff04ff17ffff04ff2fffff04ff82019fff8080808080ff8080808080808080808080ffff01ff088080ff0180ffff01ff02ffff03ffff09ff82011fff2480ffff01ff02ffff03ffff20ffff02ffff03ffff09ffff0122ffff0dff82029f8080ffff01ff02ffff03ffff09ffff0cff82029fff80ffff010280ff1080ffff01ff0101ff8080ff0180ff8080ff018080ffff01ff04ff819fffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ff82017fff8080808080808080808080ffff01ff088080ff0180ffff01ff04ff819fffff02ff3effff04ff02ffff04ff05ffff04ff0bffff04ff17ffff04ff2fffff04ff81dfffff04ff81bfffff04ff82017fff808080808080808080808080ff018080ff018080ff0180ffff01ff02ff3affff04ff02ffff04ff05ffff04ff0bffff04ff81bfffff04ffff02ffff03ff82017fffff0182017fffff01ff02ff0bffff04ff17ffff04ff2fffff01ff808080808080ff0180ff8080808080808080ff0180ff018080ffff04ffff01a0c5abea79afaa001b5427dfa0c8cf42ca6f38f5841b78f9b3c252733eb2de2726ffff04ffff0180ffff04ffff01ff02ffff01ff02ffff01ff02ffff03ff81bfffff01ff04ff82013fffff04ff80ffff04ffff02ffff03ffff22ff82013fffff20ffff09ff82013fff2f808080ffff01ff04ffff04ff10ffff04ffff0bffff02ff2effff04ff02ffff04ff09ffff04ff8205bfffff04ffff02ff3effff04ff02ffff04ffff04ff09ffff04ff82013fff1d8080ff80808080ff808080808080ff1580ff808080ffff02ff16ffff04ff02ffff04ff0bffff04ff17ffff04ff8202bfffff04ff15ff8080808080808080ffff01ff02ff16ffff04ff02ffff04ff0bffff04ff17ffff04ff8202bfffff04ff15ff8080808080808080ff0180ff80808080ffff01ff04ff2fffff01ff80ff80808080ff0180ffff04ffff01ffffff3f02ff04ff0101ffff822710ff02ff02ffff03ff05ffff01ff02ff3affff04ff02ffff04ff0dffff04ffff0bff2affff0bff2cff1480ffff0bff2affff0bff2affff0bff2cff3c80ff0980ffff0bff2aff0bffff0bff2cff8080808080ff8080808080ffff010b80ff0180ffff02ffff03ff17ffff01ff04ffff04ff10ffff04ffff0bff81a7ffff02ff3effff04ff02ffff04ffff04ff2fffff04ffff04ff05ffff04ffff05ffff14ffff12ff47ff0b80ff128080ffff04ffff04ff05ff8080ff80808080ff808080ff8080808080ff808080ffff02ff16ffff04ff02ffff04ff05ffff04ff0bffff04ff37ffff04ff2fff8080808080808080ff8080ff0180ffff0bff2affff0bff2cff1880ffff0bff2affff0bff2affff0bff2cff3c80ff0580ffff0bff2affff02ff3affff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff3effff04ff02ffff04ff09ff80808080ffff02ff3effff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01ffa07faa3253bfddd1e0decb0906b2dc6247bbc4cf608f58345d173adb63e8b47c9fffa01dd352dccae28edfcba35fbcdcab358af8d00979899dea9b8c79a9f29a1fd450a0eff07522495060c066f66f32acc2a77e3a3e737aca8baea4d1a64ea4cdc13da9ffff04ffff01a05662b49a357db4f05c2c141452b72fb91e7ec286e9b47d6c287210c63ae5cd3effff04ffff018201f4ff0180808080ffff04ffff01ff02ffff01ff02ffff01ff02ffff03ff0bffff01ff02ffff03ffff09ff05ffff1dff0bffff1effff0bff0bffff02ff06ffff04ff02ffff04ff17ff8080808080808080ffff01ff02ff17ff2f80ffff01ff088080ff0180ffff01ff04ffff04ff04ffff04ff05ffff04ffff02ff06ffff04ff02ffff04ff17ff80808080ff80808080ffff02ff17ff2f808080ff0180ffff04ffff01ff32ff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff06ffff04ff02ffff04ff09ff80808080ffff02ff06ffff04ff02ffff04ff0dff8080808080ffff01ff0bffff0101ff058080ff0180ff018080ffff04ffff01b092c7027d455e843b1226b757c1dd5e124e0e6ddab3549036a4684b410182fdfd1eee459eb76a4d6d73e94f90fbe6eb1dff018080ff018080808080ff018080808080ff01808080",
      "solution": "0xffffa0f773d07c5b0993a9c410ab5d4cb683738b40645eec9ef3ff8a605c52041e302cffa064d415895ba7e3b02c4086fd4a75b0a38199c69e6e68d636d98c5a9109b7d14dff0180ff01ffffffff80ffff01ffff81e8ffff02ffff01ff04ffff04ffff02ffff03ffff22ff27ff3780ffff01ff02ffff03ffff21ffff09ff27ffff0182616480ffff09ff27ffff01826e6d80ffff09ff27ffff0182636880ffff09ff27ffff0182747480ffff09ff27ffff0183646e7380ffff09ff27ffff0182706b80ffff09ff27ffff018272728080ffff01ff02ff02ffff04ff02ffff04ff05ffff04ff27ffff04ff37ff808080808080ffff010580ff0180ffff010580ff0180ffff04ff0bff808080ffff01ff808080ffff04ffff01ff02ffff03ff05ffff01ff02ffff03ffff09ff11ff0b80ffff01ff04ffff04ff0bff1780ff0d80ffff01ff04ff09ffff02ff02ffff04ff02ffff04ff0dffff04ff0bffff04ff17ff8080808080808080ff0180ff8080ff0180ff018080ffff826164a0d26c36cfd99da03a18a7d47dddd7beb968ff63bd7d3ccc45205fadb6958a571d80ffff33ffa07ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10ff01ffffa07ed1a136bdb4016e62922e690b897e85ee1970f1caf63c1cbe27e4e32f776d10808080ff8080808080",
    },
  ],
}
`;
